### PR TITLE
Bch replay protection

### PR DESCRIPTION
**Pull request by TheCharlatan**

### CODE REVIEW

File: src/common/utxobased/engine/makeUtxoEngine.ts [hunk 1]
```diff
@@ -241,31 +243,66 @@ export async function makeUtxoEngine(
...
+      const totalAmountToSend = spendTargets.reduce(
+        (sum, { nativeAmount }) => bs.add(sum, nativeAmount ?? '0'),
+        '0'
+      )
+      const utxos =
+        options?.utxos ?? (await processor.fetchUtxos({ utxoIds: [] }))
+
+      if (bs.gt(totalAmountToSend, `${sumUtxos(utxos)}`)) {
+        throw new InsufficientFundsError(currencyInfo.currencyCode)
+      }
```
Comment (samholmes) [inline src/common/utxobased/engine/makeUtxoEngine.ts:260]: This should be a separate PR fix; unrelated to this PR.

File: src/common/utxobased/engine/makeUtxoWalletTools.ts [hunk 2]
```diff
@@ -1,9 +1,12 @@
 import { AddressPath, CurrencyFormat, NetworkEnum } from '../../plugin/types'
+import { cdsScriptTemplates } from '../keymanager/bitcoincashUtils/checkdatasig'
```
Comment (samholmes) [inline src/common/utxobased/engine/makeUtxoWalletTools.ts:2]: Propagate the scriptTemplate in the currency's info (reference the existing bitcoin plugin architecture).

**Review by samholmes — CHANGES_REQUESTED** [review]

File: src/common/utxobased/engine/makeUtxoWalletTools.ts (hunk 2 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoWalletTools.ts]: I pushed a fixup to implement this.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 3]
```diff
@@ -881,7 +891,7 @@ const internalDeriveScriptAddress = async (
   return { address, scriptPubkey, redeemScript }
 }
 
-interface GetFreshAddressArgs extends FormatArgs { }
+interface GetFreshAddressArgs extends FormatArgs {}
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:894]: I know it's not part of the PR but on line 1234, why we are not using a const here:
``` const [tx] = await processor.fetchTransactions({ txId: utxo.txid })```

File: src/common/utxobased/info/scriptTemplates/types.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export interface ScriptTemplates {
+  [key: string]: ScriptTemplate
+}
+
+export type ScriptTemplate = (
+  pubkey: string,
+  optionalArgs?: OptionalScriptTemplateArgs
+) => string
+
+export interface OptionalScriptTemplateArgs {
```
Comment (thehobbit85) [inline src/common/utxobased/info/scriptTemplates/types.ts:10]: No need for this type here, try and keep the files outside of the info folder as "generic" as possible.

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
Minor changes, but overall, much better.

File: src/common/utxobased/info/scriptTemplates/types.ts (hunk 4 above)
Comment (TheCharlatan) [reply src/common/utxobased/info/scriptTemplates/types.ts:10]: I could not come up with a better way to define these additional parameters, that does not involve either weakening the type, or a generic that would propagate the type selection up to the core.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 3 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:894]: https://github.com/EdgeApp/edge-currency-plugins/pull/198
